Refactor `rz_core_analysis_graph`

b1llow opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

- refactor `rz_core_analysis_graph`
- move graph-related code to `cgraph.c`
- remove APIs `rz_core_print_bb_custom` `rz_core_print_bb_gml`
- fix `agl` and `ags` commands
- add unit (integration) tests for the new and reworked APIs.

**Test plan**

CI is green

**Closing issues**

Partially addresses https://github.com/rizinorg/rizin/issues/992

b1llow commented:
```shell
~/s/r/test (dev)> rizin -v
rizin 0.5.0 @ linux-x86-64
commit: 06ee5e3aaf86932789e60847c5058d7a4f04add4
~/s/r/test (dev)> rizin bins/elf/hello_world
[0x000006a0]> aa
[x] Analyze all flags starting with sym. and entry0 (aa)
[0x000006a0]> s main 
[0x000007aa]> agl
digraph code {
	graph [bgcolor=azure fontsize=8 fontname="Courier" splines="ortho"];
	node [fillcolor=gray style=filled shape=box];
	edge [arrowhead="normal"];
        "0x000007aa" -> "0x00000844" [color="#13a10e"];
        "0x000007aa" -> "0x00000806" [color="#c50f1f"];
        "0x00000806" -> "0x00000844" [color="#0037da"];
}
[0x000007aa]> ags
digraph code {
	graph [bgcolor=azure fontsize=8 fontname="Courier" splines="ortho"];
	node [fillcolor=gray style=filled shape=box];
	edge [arrowhead="normal"];
        "0x000007aa" -> "0x00000844" [color="#13a10e"];
        "0x000007aa" -> "0x00000806" [color="#c50f1f"];
        "0x00000806" -> "0x00000844" [color="#0037da"];
}
[0x000007aa]> agfd
digraph code {
	graph [bgcolor=azure fontsize=8 fontname="Courier" splines="ortho"];
	node [fillcolor=gray style=filled shape=box];
	edge [arrowhead="normal"];
	"0x000007aa" [URL="main/0x000007aa", fillcolor="white",color="#767676", fontname="Courier",label="  ; DATA XREF from entry0 @ 0x6bd\lint main (int argc, char **argv, char **envp);\l; var int64_t var_20h @ rbp-0x20\l; var int64_t var_1ch @ rbp-0x1c\l; var int64_t var_18h @ rbp-0x18\l; var int64_t var_10h @ rbp-0x10\l; var int64_t var_8h @ rbp-0x8\l0x000007aa      push  rbp\l0x000007ab      mov   rbp, rsp\l0x000007ae      sub   rsp, 0x20\l0x000007b2      lea   rax, str.Hello                                   ; 0x8d4 ; "Hello"\l0x000007b9      mov   qword [var_18h], rax\l0x000007bd      lea   rax, str.r2_folks                                ; 0x8da ; " r2-folks"\l0x000007c4      mov   qword [var_10h], rax\l0x000007c8      mov   rax, qword [var_18h]\l0x000007cc      mov   rdi, rax\l0x000007cf      call  sym.imp.strlen                                   ; size_t strlen(const char *s)\l0x000007d4      mov   dword [var_20h], eax\l0x000007d7      mov   rax, qword [var_10h]\l0x000007db      mov   rdi, rax\l0x000007de      call  sym.imp.strlen                                   ; size_t strlen(const char *s)\l0x000007e3      mov   dword [var_1ch], eax\l0x000007e6      mov   edx, dword [var_20h]\l0x000007e9      mov   eax, dword [var_1ch]\l0x000007ec      add   eax, edx\l0x000007ee      add   eax, 1\l0x000007f1      cdqe\l0x000007f3      mov   rdi, rax\l0x000007f6      call  sym.imp.malloc                                   ; void *malloc(size_t size)\l0x000007fb      mov   qword [var_8h], rax\l0x000007ff      cmp   qword [var_8h], 0\l0x00000804      je    0x844\l"]
	"0x00000844" [URL="main/0x00000844", fillcolor="white",color="#767676", fontname="Courier",label="0x00000844      nop\l0x00000845      leave\l0x00000846      ret\l"]
	"0x00000806" [URL="main/0x00000806", fillcolor="white",color="#767676", fontname="Courier",label="0x00000806      mov   rdx, qword [var_18h]\l0x0000080a      mov   rax, qword [var_8h]\l0x0000080e      mov   rsi, rdx\l0x00000811      mov   rdi, rax\l0x00000814      call  sym.imp.strcpy                                   ; char *strcpy(char *dest, const char *src)\l0x00000819      mov   rdx, qword [var_10h]\l0x0000081d      mov   rax, qword [var_8h]\l0x00000821      mov   rsi, rdx\l0x00000824      mov   rdi, rax\l0x00000827      call  sym.imp.strcat                                   ; char *strcat(char *s1, const char *s2)\l0x0000082c      mov   rax, qword [var_8h]\l0x00000830      mov   rdi, rax\l0x00000833      call  sym.imp.puts                                     ; int puts(const char *s)\l0x00000838      mov   rax, qword [var_8h]\l0x0000083c      mov   rdi, rax\l0x0000083f      call  sym.imp.free                                     ; void free(void *ptr)\l"]
        "0x000007aa" -> "0x00000844" [color="#13a10e"];
        "0x000007aa" -> "0x00000806" [color="#c50f1f"];
        "0x00000806" -> "0x00000844" [color="#0037da"];
}
```

before this PR, `ag[sl]` doesn't get more information than `agfd`, maybe they should be removed.

@XVilka what do you think ?

XVilka commented on librz/core/agraph.c:
```diff
@@ -4321,7 +4321,7 @@ RZ_API int rz_core_visual_graph(RzCore *core, RzAGraph *g, RzAnalysisFunction *_
 			if (fcn && rz_cons_yesno('y', "Compute function callgraph? (Y/n)")) {
 				rz_core_agraph_reset(core);
-				rz_core_cmd0(core, ".agc* @$FB;.axfg @$FB");
+				rz_core_cmd0(core, ".agc * @$FB;.axfg @$FB");
```
Is it possible to convert to the API while you are on it?

b1llow replied:
Yes but needs some more work

b1llow replied:
But I found that `axfg` seems to be an API that has been removed

XVilka commented on librz/core/canalysis.c:
```diff
@@ -1025,7 +1025,7 @@ static int __core_analysis_fcn(RzCore *core, ut64 at, ut64 from, int reftype, in
 }
 
-static char *get_title(ut64 addr) {
+RZ_API char *rz_address_str(ut64 addr) {
```
No need for this as an API - in fact, I even recommend inlining these calls.

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,384 @@
+// SPDX-FileCopyrightText: 2022 imbillow <billow.fun@gmail.com>
```
Have you rewritten *everything* from scratch? If not - please preserve also the original copyrights from `canalysis.c`.

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,384 @@
+	goto ret;
+}
+
+RZ_API RZ_OWN RzGraph /*<RzGraphNodeInfo *>*/ *rz_core_graph_function(RzCore *core, ut64 addr) {
```
Please add Doxygen for those.

XVilka commented on librz/include/rz_core.h:
```diff
@@ -856,6 +804,42 @@ RZ_API RZ_BORROW const char *rz_core_analysis_name_type_to_str(RzCoreAnalysisNam
+RZ_API char *rz_address_str(ut64 addr);
+
+/* cgraph.c */
+typedef enum {
```
I recommend adding brief Doxygen description of these structures - what are they for

XVilka reviewed (changes requested):
Overall looks good.

XVilka commented:
@imbillow, judging by the output `ag{s,l}` seem to draw only lines. Sometimes it's useful, I would keep it. More concerning that they both return the same output. Have you experimented with other functions? Maybe the difference would be noticeable on more complex graphs? Anyway, it's better to check the difference in the code.
I don't see `ags` in `librz/core/cmd_descs/cmd_analysis.yml` - maybe remove it? Why it does even work?

b1llow committed:
Refactor `rz_core_analysis_graph*()` functions

- fix `ags` and `agl` commands
- add more tests for the `ag` commands
- move graph related code to cgraph.c
- remove APIs `rz_core_print_bb_custom` `rz_core_print_bb_gml`
- exports
    - `rz_graph_drawable_to_json_str`
    - `rz_graph_drawable_to_cmd`
    - `rz_graph_drawable_to_gml`
    - `rz_core_graph`
    - `rz_core_graph_diff`
    - `rz_core_graph_format_from_string`
    - `rz_core_graph_type_from_string`
    - `rz_core_graph_write`
    - `create_agraph_from_graph_at`
- add unit/integration tests for the graph API

b1llow commented:
OK

b1llow commented:
> @imbillow, judging by the output `ag{s,l}` seem to draw only lines. Sometimes it's useful, I would keep it. More concerning that they both return the same output. Have you experimented with other functions? Maybe the difference would be noticeable on more complex graphs? Anyway, it's better to check the difference in the code. I don't see `ags` in `librz/core/cmd_descs/cmd_analysis.yml` - maybe remove it? Why it does even work?

Before converting `ag[sl]` to `rzshell` there is no documentation about them and even `ag?` will not have their help text.
It's just that they do have their processing logic in the code.

XVilka commented on librz/core/agraph.c:
```diff
@@ -4319,9 +4319,21 @@ RZ_API int rz_core_visual_graph(RzCore *core, RzAGraph *g, RzAnalysisFunction *_
-				rz_core_cmd0(core, ".agc* @$FB;.axfg @$FB");
+				rz_core_agraph_apply(core, graph);
+				rz_graph_free(graph);
+				rz_core_cmd0(core, ".axfg @$FB");
```
Is there a chance to convert this one too? Can be done in a separate PR, though.

b1llow replied:
It's just that the current version of rizin does not have this command.

![image](https://user-images.githubusercontent.com/24983369/186744985-2fb13649-267b-491c-bc69-90e3ca8b528e.png)

XVilka replied:
Well, the `rz_core_cmd0()` wouldn't work either in that case obviously, so it has to be fixed somehow.

XVilka commented on librz/core/agraph.c:
```diff
@@ -4908,42 +4920,36 @@ RZ_API int rz_core_visual_graph(RzCore *core, RzAGraph *g, RzAnalysisFunction *_
 
 /**
- * @brief Create RzAGraph from generic RzGraph with RzGraphNodeInfo as node data
+ * @brief Create RzAGraph from generic RzGraph with RzGraphNodeInfo as node data at \p ag from \p g
```
Use `\brief` syntax please

XVilka commented on librz/core/agraph.c:
```diff
@@ -4908,42 +4920,36 @@ RZ_API int rz_core_visual_graph(RzCore *core, RzAGraph *g, RzAnalysisFunction *_
+ * @brief Create RzAGraph from generic RzGraph with RzGraphNodeInfo as node data at \p ag from \p g
  *
- * @param graph <RzGraphNodeInfo>
  * @return RzAGraph* NULL if failure
```
`\return`

XVilka commented on librz/core/agraph.c:
```diff
@@ -4908,42 +4920,36 @@ RZ_API int rz_core_visual_graph(RzCore *core, RzAGraph *g, RzAnalysisFunction *_
-		return NULL;
-	}
-	result_agraph->need_reload_nodes = false;
+RZ_API RzAGraph *create_agraph_from_graph_at(RZ_NONNULL RzAGraph *ag, RZ_NONNULL const RzGraph /*<RzGraphNodeInfo *>*/ *g) {
```
Add ownership, please.

XVilka commented on librz/core/agraph.c:
```diff
@@ -4956,14 +4962,30 @@ RZ_API RzAGraph *create_agraph_from_graph(const RzGraph /*<RzGraphNodeInfo *>*/
 }
+
+/**
+ * @brief Create RzAGraph from generic RzGraph with RzGraphNodeInfo as node data
```
Use `\brief` syntax

XVilka commented on librz/core/agraph.c:
```diff
@@ -4956,14 +4962,30 @@ RZ_API RzAGraph *create_agraph_from_graph(const RzGraph /*<RzGraphNodeInfo *>*/
+ * @param graph <RzGraphNodeInfo>
+ * @return RzAGraph* NULL if failure
+ */
+RZ_API RzAGraph *create_agraph_from_graph(const RzGraph /*<RzGraphNodeInfo *>*/ *graph) {
```
The same concern about the API. In fact, looks like only this one should be an API while the previous could be a `static` function.

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,726 @@
+	RzList *calls = rz_core_analysis_fcn_get_calls(core, fcn);
+	RzListIter *it;
+	RzAnalysisXRef *xref;
+	// TODO: maybe fcni->calls instead ?
```
Remove this TODO - it's unnecessary.

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,726 @@
+	}
+}
+
+static inline RzGraph /*<RzGraphNodeInfo *>*/ *rz_core_graph_function_bbs(RzCore *core, ut64 addr, GraphBodyFn body_fn) {
```
No need for `inline` here. Moreover, while it's static, it still deserves a short Doxygen comment about the purpose - it's unclear from the function name itself.

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,726 @@
+	return RZ_CORE_GRAPH_TYPE_UNK;
+}
+
+RZ_IPI bool rz_core_graph_print_graph(RzCore *core, RzGraph /*<RzGraphNodeInfo *>*/ *graph, RzCoreGraphFormat format, int use_utf, bool use_offset, bool is_callgraph) {
```
These arguments probably can be combined into a structure instead.

XVilka commented on librz/core/cmd/cmd_analysis.c:
```diff
@@ -5672,175 +5269,91 @@ RZ_IPI char **rz_analysis_graph_format_choices(RzCore *core) {
 
 RZ_IPI RzCmdStatus rz_analysis_graph_normal_handler(RzCore *core, int argc, const char **argv) {
-	rz_core_analysis_graph(core, core->offset, 0);
+	//	RzGraph *graph = rz_core_graph_function_bbs(core, core->offset);
```
Forgotten?

XVilka commented on librz/core/core_private.h:
```diff
@@ -214,4 +218,8 @@ static inline RzCmdStatus bool2status(bool val) {
 	return val ? RZ_CMD_STATUS_OK : RZ_CMD_STATUS_ERROR;
 }
 
+static inline char *rz_address_str(ut64 addr) {
```
Unnecessary?

XVilka commented on librz/include/rz_core.h:
```diff
@@ -857,6 +805,55 @@ RZ_API void rz_core_analysis_name_free(RZ_NULLABLE RzCoreAnalysisName *p);
+} RzCoreGraphFormat;
+
+/**
+ * \brief RzGraph type, see ag?
```
Referring to the `ag?` here is a bad idea - we could change the command, and it will be completely out of sync. I recommend to put full description here, after all Rizin is also a library, not just a tool

XVilka commented on librz/util/graph_drawable.c:
```diff
@@ -70,21 +72,19 @@ RZ_API char *rz_graph_drawable_to_dot(RzGraph /*<RzGraphNodeInfo *>*/ *graph, co
 }
 
-RZ_API void rz_graph_drawable_to_json(RzGraph /*<RzGraphNodeInfo *>*/ *graph, PJ *pj, bool use_offset) {
+RZ_API void rz_graph_drawable_to_json(RZ_NONNULL RzGraph /*<RzGraphNodeInfo *>*/ *graph, RZ_NONNULL PJ *pj, bool use_offset) {
```
Please add a brief description

XVilka commented on librz/util/graph_drawable.c:
```diff
@@ -98,11 +98,97 @@ RZ_API void rz_graph_drawable_to_json(RzGraph /*<RzGraphNodeInfo *>*/ *graph, PJ
 	pj_end(pj);
 }
+
+RZ_API RZ_OWN char *rz_graph_drawable_to_json_str(RZ_NONNULL RzGraph /*<RzGraphNodeInfo *>*/ *graph, bool use_offset) {
```
Same.

XVilka commented on librz/util/graph_drawable.c:
```diff
@@ -98,11 +98,97 @@ RZ_API void rz_graph_drawable_to_json(RzGraph /*<RzGraphNodeInfo *>*/ *graph, PJ
+	return rz_strbuf_drain(sb);
+}
+
+RZ_API RZ_OWN char *rz_graph_drawable_to_cmd(RZ_NONNULL RzGraph /*<RzGraphNodeInfo *>*/ *graph) {
```
And here

XVilka commented on librz/util/graph_drawable.c:
```diff
@@ -98,11 +98,97 @@ RZ_API void rz_graph_drawable_to_json(RzGraph /*<RzGraphNodeInfo *>*/ *graph, PJ
+	return rz_strbuf_drain(sb);
+}
+
+RZ_API RZ_OWN char *rz_graph_drawable_to_gml(RZ_NONNULL RzGraph /*<RzGraphNodeInfo *>*/ *graph) {
```
You got the idea

XVilka commented on test/db/cmd/cmd_ag:
```diff
@@ -783,7 +772,7 @@ agraph.color_box=G1swbQ==
 agraph.delta_y=0x1
 agraph.h=0xe
-agraph.is_callgraph=false
+agraph.is_callgraph=true
```
Why this change?

b1llow replied:
It seems that is_callgraph controls whether the edges of the ascii graph are linked with `'>'` or `'t' | 'f'`

`ag[cC]` as Function callgraph, which I think is the correct behavior

XVilka commented on test/db/cmd/cmd_ag:
```diff
@@ -920,7 +909,6 @@ RUN
 
 NAME=agf multiple formats
 FILE=bins/elf/hello_world
-BROKEN=1
```
So cool! 👍

b1llow replied:
yeah

XVilka reviewed (changes requested):
Quite nice cleanup.

b1llow commented:
I found that the tests here seem to fail randomly.

```
[XX] db/archos/darwin-arm64/dbg dcu + read
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-1245088262/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -d -Qc 'dcu main
pi 8 @ pc
' bins/mach0/hello-macos-arm64
-- stderr
--- expected
+++ actual
@@ -1,2 +1,4 @@
+exec: No such file or directory
+exec: No such file or directory
 Continue until c88
 hit breakpoint at: c88
```

https://ci.rizin.re/rizinorg/rizin/build/851/6

wargio commented on librz/core/cagraph.c:
```diff
@@ -131,17 +131,21 @@ static void agraph_print_edge(RzANode *from, RzANode *to, void *user) {
-	encbody = rz_base64_encode_dyn((const ut8 *)n->body, len);
-	cmd = rz_str_newf("agn \"%s\" base64:%s\n", n->title, encbody);
+	if (RZ_STR_ISEMPTY(n->body)) {
+		cmd = rz_str_newf("agn \"%s\"\n", n->title);
+	} else {
+		char *encbody = rz_base64_encode_dyn((const ut8 *)n->body, len);
+		cmd = rz_str_newf("agn \"%s\" base64:%s\n", n->title, encbody);
```
this looks like a command injection bug. maybe is worty to open an issue?

b1llow replied:
yes

XVilka commented:
@imbillow as you are working on the API, I also recommend creating a new integration test in [`test/integration/`](https://github.com/rizinorg/rizin/tree/dev/test/integration) that would open some binary, analyzes it, check CFG, callgraph, references graphs on a few selected (better with more nodes) functions from that binary. Moreover, it makes sense also to save the project, close it, load again, and check if the graphs were properly stored/matched again exactly. As this is a core feature of any RE framework, it should be better tested than now. See [`test/integration/test_open_analyse_save_load_project.c`](https://github.com/rizinorg/rizin/blob/dev/test/integration/test_open_analyse_save_load_project.c) as an example.

XVilka commented:
@imbillow, please add tests for `agl` and `ags` too, to avoid future breaks.

XVilka commented:
Please also fix tests for the new output:
```
[XX] db/cmd/cmd_graph agg one
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'agn 0x0 base64:cHVzaCBlYnAKbW92IGVicCwgZXNwCnBvcCBlYnAKcmV0
agg
' =
-- stdout
--- expected
+++ actual
@@ -1,7 +1,7 @@
-.--------------------.
-|  0x0               |
-| push ebp           |
-| mov ebp, esp       |
-| pop ebp            |
-| ret                |
-`--------------------'
+.--------------.
+|  0x0         |
+| push ebp     |
+| mov ebp, esp |
+| pop ebp      |
+| ret          |
+`--------------'
```

b1llow commented:
I want to write a script to fix tests automatically so it might be a bit late.

XVilka commented:
@imbillow no need for script, just use `-i` option of the `rz-test`. It will ask you if the new output is better and update the test.

b1llow commented:
@XVilka OK, thank you. I didn't know that before.

XVilka commented:
@imbillow now that PR is close to its final, mergeable form, thanks. Just please don't forget https://github.com/rizinorg/rizin/pull/2949#issuecomment-1227896945

XVilka commented on librz/core/agraph.c:
```diff
@@ -3762,6 +3757,18 @@ RZ_API RzANode *rz_agraph_add_node_with_color(const RzAGraph *g, const char *tit
+/**
+ * \brief Convert a RzGraphNodeInfo \p info to RzANode and add to \p g.
+ */
+RZ_API RzANode *rz_agraph_add_node_from_node_info(RZ_NONNULL const RzAGraph *g, RZ_NONNULL const RzGraphNodeInfo *info) {
```
If you add `RZ_NONNULL`, please add also `rz_return_val_if_fail`, please also ownership mark on the return value.

XVilka commented on librz/core/agraph.c:
```diff
@@ -4319,10 +4326,23 @@ RZ_API int rz_core_visual_graph(RzCore *core, RzAGraph *g, RzAnalysisFunction *_
-				rz_core_cmd0(core, ".agc* @$FB;.axfg @$FB");
-				rz_core_agraph_print_interactive(core);
+				if (rz_core_agraph_apply(core, graph)) {
+					rz_core_cmd0(core, ".axfg @$FB");
```
Please add here `// TODO: Convert to the API`

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,755 @@
+			}
+		}
+
+		if (bbi->switch_op) {
```
Maybe can be reversed, `if (!bbi->switch_op) { continue }`

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,755 @@
+/**
+ * \brief Get a graph of the function blocks at \p addr.
+ */
+RZ_API RZ_OWN RzGraph /*<RzGraphNodeInfo *>*/ *rz_core_graph_function(RZ_NONNULL RzCore *core, ut64 addr) {
```
RZ_NONNULL for core here is unnecessary, it's implied

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,755 @@
+/**
+ * \brief Get a graph of the function difference at \p addr.
+ */
+RZ_API RZ_OWN RzGraph /*<RzGraphNodeInfo *>*/ *rz_core_graph_diff(RZ_NONNULL RzCore *core, ut64 addr) {
```
RZ_NONNULL for core here is unnecessary, it's implied

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,755 @@
+/**
+ * \brief Get a line graph of the function blocks at \p addr.
+ */
+RZ_API RZ_OWN RzGraph /*<RzGraphNodeInfo *>*/ *rz_core_graph_line(RZ_NONNULL RzCore *core, ut64 addr) {
```
RZ_NONNULL for core here is unnecessary, it's implied

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,755 @@
+/**
+ * \brief Get a normal graph of the function blocks at \p addr.
+ */
+RZ_API RZ_OWN RzGraph /*<RzGraphNodeInfo *>*/ *rz_core_graph_normal(RZ_NONNULL RzCore *core, ut64 addr) {
```
RZ_NONNULL for core here is unnecessary, it's implied

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,755 @@
+/**
+ * \brief Get a graph of specific type (\p type) at \p addr.
+ */
+RZ_API RZ_OWN RzGraph /*<RzGraphNodeInfo *>*/ *rz_core_graph(RZ_NONNULL RzCore *core, RzCoreGraphType type, ut64 addr) {
```
RZ_NONNULL for core here is unnecessary, it's implied

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,755 @@
+ * \brief Convert a string to RzCoreGraphType.
+ */
+RZ_API RzCoreGraphType rz_core_graph_type_from_string(RZ_NULLABLE const char *x) {
+	rz_return_val_if_fail(x, RZ_CORE_GRAPH_TYPE_UNK);
```
You say "nullable" but have an assert, remove either, but don't combine together

XVilka commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,755 @@
+	return RZ_CORE_GRAPH_TYPE_UNK;
+}
+
+RZ_IPI bool rz_core_graph_print_graph(RzCore *core, RzGraph /*<RzGraphNodeInfo *>*/ *graph, RzCoreGraphFormat format, bool use_offset) {
```
I suppose `graph` here shouldn't be NULL, right? Please add a mark and check

XVilka commented on test/integration/test_analysis_graph.c:
```diff
@@ -0,0 +1,92 @@
+    rz_core_analysis_everything(core, false, "esil");
+
+    // 3. dataref graph
+    RzGraph *g = graph_by_function_name(core, "entry0");
```
Can you use some more complex functions than with just one node? Also, maybe multiple graph types at once: CFG, callgraph, refs.

XVilka reviewed (changes requested):
Overall looks good, a few nitpicks remain.

XVilka commented on librz/core/cgraph.c:
```diff
@@ -536,7 +537,6 @@ RZ_API RzCoreGraphFormat rz_core_graph_format_from_string(RZ_NULLABLE const char
  */
 RZ_API RzCoreGraphType rz_core_graph_type_from_string(RZ_NULLABLE const char *x) {
-	rz_return_val_if_fail(x, RZ_CORE_GRAPH_TYPE_UNK);
 	if (strcmp(x, "dataref") == 0) {
```
I think `strcmp(NULL)` would crash. You probably want to check for NULL before calling it.

XVilka reviewed (approved):
LGTM. Let's wait for more approvals first though.

XVilka commented:
@imbillow I just merged removing the eprintfs PR, there are conflicts now, could you please rebase your PR on top of the `dev` once more?

wargio commented on librz/core/cagraph.c:
```diff
@@ -201,3 +205,80 @@ RZ_IPI void rz_core_agraph_print_gml(RzCore *core) {
+	case RZ_CORE_GRAPH_FORMAT_CMD:
+		rz_core_agraph_print_rizin(core);
+		break;
+	case RZ_CORE_GRAPH_FORMAT_JSON:
```
add a `/* fall-thru */` here

wargio commented on librz/core/cgraph.c:
```diff
@@ -0,0 +1,760 @@
+		string = rz_core_graph_to_dot_str(core, graph);
+		break;
+	}
+	case RZ_CORE_GRAPH_FORMAT_JSON:
```
add a `/* fall-thru */` here

XVilka commented:
@imbillow please address the comments, squash all commits into one with a proper commit message and let's merge it once green.

XVilka commented:
@imbillow no worries, did it myself.
